logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueReader.java

```diff
@@ -101,8 +101,18 @@ public void seekToNextEvent(Timestamp timestamp) throws IOException {
         }
     }
 
+    /**
+     * Opens the segment reader for the given path.
+     * Side effect: Will attempt to remove the given segment from the list of active
+     *              segments if segment is not found.
+     * @param segment Path to segment File
+     * @return Optional containing a RecordIOReader if the segment exists
+     * @throws IOException
+     */
     private Optional<RecordIOReader> openSegmentReader(Path segment) throws IOException {
         if (!Files.exists(segment)) {
+            // file was deleted by upstream process and segments list wasn't yet updated
+            segments.remove(segment);
```

Comment:
Placing this remove line in L88 condition would be better for me. Because `openSegmentReader` method now holding the hidden logic and we cannot figure out unless we read the method description. Also moving method description to public `seekToNextEvent` would really help when going through the high level source logic.

Reply:
If we move the remove of segment from `segments` set to line 88 we have to do it for each call of `openSegmentReader`, because this method already handle the consistency of `segments` field.
Maybe we could rename the method to better express the logic it applies.
WDYT?

Reply:
The problem is that `openSegmentReader` is called in four separate places, each of which, could potentially trigger this issue - this particular hang bug was in a different place to L88, and leaving the clean up as an exercise for the caller has dangers.

I agree that side effects in functions are unfortunate, and toyed with the idea of renaming the method to something like `cleaningOpenSegmentReader`

Reply:
Thanks for explanations. I agree with @robbavey to give descriptive name when we have a chance to update the class.